Implement a CLI option to set particular CPU cores for BenchExec

IljaZakharov opened the pull request:
This adds an option to provide the set of particular CPU cores identifiers to BenchExec. The option implements separating available CPU cores for independent BenchExec instances to avoid usage of the same CPUs.

PhilippWendler commented on benchexec/benchexec.py:
```diff
@@ -152,6 +152,13 @@ def create_argument_parser(self):
                           metavar="N",
                           help="Limit each run of the tool to N CPU cores (-1 to disable).")
 
+        parser.add_argument("-s", "--allowedCores", dest="coreset",
```
I would prefer going without a short option because this option is not one that will be used by many people (I guess).

PhilippWendler commented on benchexec/benchexec.py:
```diff
@@ -152,6 +152,13 @@ def create_argument_parser(self):
+                            metavar="n",
+                            nargs='+',
+                            type=int,
+                            help="Limit each run of the tool to a subset of particular CPU cores.")
```
Can you write "Limit the set of cores BenchExec will use for all runs" instead, please? The current description sounds like one run will be able to use all the given cores.

PhilippWendler commented on benchexec/benchexec.py:
```diff
@@ -152,6 +152,13 @@ def create_argument_parser(self):
+        parser.add_argument("-s", "--allowedCores", dest="coreset",
+                            default=None,
+                            metavar="n",
+                            nargs='+',
```
Instead of using `nargs`, I would suggest making it possible to give a `1-3,5` style list. We have an existing utility method `parse_int_list` for this.

PhilippWendler commented on benchexec/resources.py:
```diff
@@ -68,11 +68,20 @@ def get_cpu_cores_per_run(coreLimit, num_of_threads, my_cgroups):
+
+        # Filter CPU cores according to the list of identifiers provided by a user
+        if coreSet:
+            invalid_cores = list(map(str, sorted(set(coreSet).difference(set(allCpus)))))
```
Why mapping to `str` here?

PhilippWendler commented on benchexec/resources.py:
```diff
@@ -68,11 +68,20 @@ def get_cpu_cores_per_run(coreLimit, num_of_threads, my_cgroups):
 
     @param coreLimit: the number of cores for each run
     @param num_of_threads: the number of parallel benchmark executions
+    @param coreSet: the list of CPU cores identifiers provided by a user
```
Please mention that `None` makes benchexec use all cores.

IljaZakharov commented:
Changed option a bit as you requested and removed the redundant cast. Found out that the option works only if the CPU core limit is also provided (added the notion of this to the help message).

PhilippWendler commented on benchexec/benchexec.py:
```diff
@@ -152,6 +152,12 @@ def create_argument_parser(self):
+        parser.add_argument("--allowedCores",
+                          dest="coreset", default=None, type=util.parse_int_list,
+                          help="Limit the set of cores BenchExec will use for all runs "
+                               "(Applied only if the number of CPU cores is limited).",
```
I actually like this because it means that most users will keep using the normal option for the core limit instead of using this "manual" option. But could you please change it such that benchexec aborts with an invalid-configuration error if this option is given without a core limit? Silently ignoring this option seems very dangerous.

IljaZakharov commented:
As BenchExec uses the option if the limitation of cores is provided either within XML or CLI option, I added termination only if both are missing.
